File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/RefreshDataWorker.java
package com.solutions.alphil.zambiajobalerts.classes;

import android.content.Context;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;

public class RefreshDataWorker extends Worker {
    public RefreshDataWorker(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);
    }

    @NonNull
    @Override
    public Result doWork() {
        Log.d("RefreshDataWorker", "Starting background sync...");
        
        // 1. Refresh Job Data
        JobRepository repository = new JobRepository((android.app.Application) getApplicationContext());
        repository.refreshJobs();
        
        // 2. Refresh/Pre-cache Ads
        // AdManager dispatches the Mobile Ads request onto the main thread when this worker runs in the background.
        try {
            AdManager.getInstance().loadInterstitialAd(getApplicationContext());
        } catch (Exception e) {
            Log.e("RefreshDataWorker", "Error pre-caching ads: " + e.getMessage());
        }

        return Result.success();
    }
}

File: app/src/main/java/com/solutions/alphil/zambiajobalerts/classes/RefreshDataWorker.kt
package com.solutions.alphil.zambiajobalerts.classes

import android.app.Application
import android.content.Context
import android.util.Log
import androidx.work.Worker
import androidx.work.WorkerParameters

class RefreshDataWorker(context: Context, workerParams: WorkerParameters) : Worker(context, workerParams) {
    override fun doWork(): Result {
        Log.d(TAG, "Starting background sync...")

        val app = applicationContext as? Application
        if (app != null) {
            JobRepository(app).refreshJobs()
        } else {
            Log.w(TAG, "Application context was not an Application; skipping job refresh")
        }

        try {
            AdManager.getInstance().loadInterstitialAd(applicationContext)
        } catch (error: Exception) {
            Log.e(TAG, "Error pre-caching ads: ${error.message}")
        }

        return Result.success()
    }

    companion object {
        private const val TAG = "RefreshDataWorker"
    }
}
